Gpio and esp32 support

hetii opened the pull request:
Hello dear.

This is my first PR for your project.

The main changes include adding dynamic GPIO ports support (based on yaml configuration) with frontend visualization.

Adding information modals powered by REST endpoints including handling uart settings over web ui.

Stability fixes and support for ESP32 - no more odd crashes.

I have resolved almost all "welcome PRs" from your README.md (except for macOS support).

Below are a few screenshots:
<img width="2559" height="318" alt="intro1" src="https://github.com/user-attachments/assets/a547bf63-52e8-4229-8bc4-73fd7757a8de" />

<img width="627" height="770" alt="intro2" src="https://github.com/user-attachments/assets/09971532-3099-49be-9d5a-adc0b94a31f5" />

<img width="617" height="708" alt="intro5" src="https://github.com/user-attachments/assets/b946a990-1038-4ac9-b8d1-ae9fa9bf0a59" />

<img width="616" height="329" alt="intro3" src="https://github.com/user-attachments/assets/2d4c68a0-381b-4c23-8932-55b05063c6e2" />

After merging, this version should be tagged as xterm-4.9 because,
the second PR will concern raising all packages in the frontend, including xterm.

Things to note:

My code compiles using the latest versions of ESPAsyncWebServer, and I saw that when you created this project, you modified the library available at the time.

I'm not sure if your patches are still required with the current versions of ESPAsyncWebServer, so if you set debug.legacy_lib to true, the project will still use your changes.

The basic UART port support for esp32 works, but I haven't tested much whether the mechanisms contained in src/ExtendedSerial.cpp such as autobaudMeasure(), sendBreak(), etc. work correctly.

I hope that my PRs will add value to your project.

Best regards.

hetii committed:
Created workaround to prevent bootloop when browser call /reset endpoint.

hetii committed:
Allow controling CORS by config file.

hetii committed:
Add http service to mdns also when OTA is disabled.

hetii committed:
Use optional hardware to disable bootlog.

hetii committed:
Remove auto generated files.

hetii committed:
Improve handling websocket in case of race condition during client disconnection.

hetii committed:
Add gpio control for backend.

hetii committed:
Add build type and exception decoder to debug section.

hetii committed:
Set the default button colors if missing in yaml.

hetii committed:
Add gpio control for frontend.

hetii committed:
Add extra information to readme and config file.

hetii committed:
Add CTRL+` as an alias for CTRL+Q.

hetii committed:
Add browser issues to troubleshooting section.

hetii committed:
Fix wrapper style for terminal and header bar.

hetii committed:
Fix frontend build description.

hetii committed:
Fix object keys for generated leds and buttons.

hetii committed:
Fix syntax and do optimization to satisfy linter.

hetii committed:
Add fakeesp.js test server.

hetii committed:
Add missing information for info-modal.

hetii committed:
Add support for es2018 typescript target.

hetii committed:
Fix crashes on websocket calls during initial messages.

hetii committed:
Fix build when debug is enabled.

hetii committed:
Cleanups

hetii committed:
Add support for ESP32.

hetii committed:
Don't set by default adc_input and tx_en in config example.

hetii commented:
Hello @depau.
I'm glad that you are happy with my contribution.
As you asked, here is my second PR: https://github.com/hetii/wi-se-sw/pull/1

I upgrade most packages and dependencies, so the final html.h grow about 196KB, but just 3.2% in flash after compilation:
```
xterm 5.5:
size of include/html.h: 908551 (888K)
RAM:   [====      ]  39.9% (used 32708 bytes from 81920 bytes)
Flash: [======    ]  57.9% (used 604275 bytes from 1044464 bytes)

xterm 4.9:
size of include/html.h: 708567 (692K)
RAM:   [====      ]  39.9% (used 32708 bytes from 81920 bytes)
Flash: [=====     ]  54.7% (used 571523 bytes from 1044464 bytes)
```

Regards.

depau reviewed (changes requested):
Some early feedback since I noticed your last commit renaming `/ws` to `/wisews`: this actually breaks https://github.com/depau/ttyc/ which I regularly use.

Is the change necessary or can we find a workaround for the issue?

In general I'd like to keep the protocol compatible with [ttyd](https://tsl0922.github.io/ttyd/), making only additive changes for Wi-Se specific functionality.

hetii commented:
@depau 
Don't worry, take your time.

I feel like I need to spend a little more time refining my branch because I found some issues with GPIO handling.

Even when I use digitalWrite() to turn the GPIO on for a specific time, it seems to work more like analogWrite(),
e.g., when I turn it on for 1 second, my LED shines dimmer than when I do the same but for 10 seconds. 
It's a bit strange, but it works the same on esp8266 and esp32, so it needs to be fixed.

As for /ws vs /wisews, it turns out that the old version of webpack -dev-server uses the /websocket endpoint for HPM and reload functionality, so there is no conflict when using a proxy, but the new version also uses /ws, which leads to a collision because the proxy tries to redirect its websocket to the target endpoint, resulting in a kind of reconnection loop in the console.

I believe your compatibility requirements with ttyd are more important, so I will try to find a solution to this problem as well.

hetii committed:
Fix gpio handler.

hetii commented:
Hello again @depau.

I fixed gpio and websocket issues in new webpack-dev-server.

To avoid the mess I made in my commits,
I rewrote the history of my branches and created a new PR hetii#4 with the new xterm and webpack.

Now (I hope) everything should be fine.

Waiting for your feedback, regards.

depau commented on configs/config.yml.example:
```diff
@@ -34,9 +40,64 @@ board:
+    #   inverted: true
+    # # Output to reset target board, eg, moterboard reset button.
+    # - gpio: 12
+    #   mode: OUTPUT
```
nitpick: typo: moterboard → motherboard

depau commented on builder/platformio.j2.ini:
```diff
@@ -38,6 +47,16 @@ upload_flags =
     --auth={{ cfg.ota_password }}
 {% endif %}
 lib_deps =
-	ESP Async WebServer=https://github.com/Depau/ESPAsyncWebServer/archive/wi-se-patches.zip
+{% if cfg.build_legacy_lib %}
+	ESPAsyncWebServer=https://github.com/Depau/ESPAsyncWebServer/archive/wi-se-patches.zip
+{% else %}
+    ESP32Async/ESPAsyncWebServer
+{% endif %}
+{% if cfg.board_mcu == 'esp8266' %}
 	ESP8266WiFi
+{% endif %}
 	ArduinoJson
+
+build_type = {{ cfg.build_type }}
+monitor_speed = {{ cfg.serial_debug_baud }}
+monitor_filters = {{ cfg.exception_decoder }}
```
(no action required)

I think you made the right choices when adding ESP32 support given the previous status, however I'm not really happy about my previous choices that led to this :)

When I designed this config generator I really didn't know any better but honestly... I think it kinda sucks. It's nice it allows doing OTA upgrades on the entire fleet but it's quite unnecessarily bespoke.

Do you think we could do any better? This could have been multiple PlatformIO profiles, and I think we can figure something out to launch a generator for the config header.

What do you think?

hetii replied:
Exactly as you say, I didn't want to introduce any new revolutions, so I just based my work on what I found.
If there were to be any changes here, it would be later in the project's life cycle, after the current changes have been approved.

depau commented on configs/config.yml.example:
```diff
@@ -74,6 +134,18 @@ uart:
     baud: 115200
     #config: '(UART_NB_BIT_8 | UART_PARITY_NONE | UART_NB_STOP_BIT_1)'
 
+    # Additional gpio that can be used to disable debug output at boot time on tx line of serial 0.
+    # Below hardware is required:
+    # https://github.com/nopnop2002/esp8266-mqtt-uart-bridge?tab=readme-ov-file#add-a-circuit
+    #                 2 x Schottky diodes 1N5819 or BAS85 (1n4148 won't work!)
+    # 3v3 ---[1k]---,       /
+    # tx_en --------*---(>|)---,
+    # gnd ---[1k]--------------|
+    # esp8266_tx -------(>|)---*---> RX_TARGET
+    #
+    # NOTE: GPIO15 is pulled down and cannot be used for this.
+    # tx_en: 5
```
This is really nice!

hetii replied:
Yep, but type of schottky diodes is important for the way how it works.

depau commented on html/README.md:
```diff
@@ -4,7 +4,9 @@ https://github.com/tsl0922/ttyd/
 ## Prerequisites
 
-install [Yarn](https://yarnpkg.com), and run: `yarn install`.
+Install [Node.js](https://nodejs.org/en) v15.14.0, eg: `nvm install v15.14.0`.
```
Isn't Node 15.14 a bit old? What's stopping us from going with the current version?

depau replied:
Oh I see, it's in the other PR. Nevermind!

depau commented on include/server.h:
```diff
@@ -110,7 +115,11 @@ class WiSeServer {
             panic();
+#else
+            esp_system_abort("Critical error, all client command caches full!");
+#endif
```
nitpick: I see you created `compat.h` for this purpose, can this be turned into a `wisePanic(whatever_string_t)` helper and the ifdef moved there?

hetii replied:
Done

depau commented on src/server.cpp:
```diff
@@ -33,15 +38,26 @@ void WiSeServer::begin() {
+              std::bind(&WiSeServer::handleGpioRequest, this, std::placeholders::_1),
+              nullptr,
+              std::bind(&WiSeServer::handleGpioBody, this, std::placeholders::_1, std::placeholders::_2,
+                        std::placeholders::_3, std::placeholders::_4, std::placeholders::_5));
```
nitpick towards myself: when I wrote this code I barely new C++, but some time after I worked for a few years on large C++ projects at work so I got to know the language better and... I think this placeholder business is dumb. I should probably see if we can switch to a newer C++ version, do things better and remove this eyesore.

depau commented on src/server.cpp:
```diff
@@ -33,15 +38,26 @@ void WiSeServer::begin() {
+        // Schedule reboot and let clients know about it.
+        this->shouldReboot = millis() + 1000;
+        end();
+        // Don't do it here - check loop() in main.cpp for details.
+        // ESP.reset();
```
nitpick: You can remove this comment, though thanks for guiding me through this change.

hetii replied:
Yep I just like to left some comment for historical reason to have a context, but I will remove it like you request.

depau commented on src/server.cpp:
```diff
@@ -316,6 +325,81 @@ WiSeServer::handleSttyBody(
+    for (size_t i = 0; i < TARGET_GPIO_COUNT; i++) {
+        // OUTPUTs
+        if ((gpioConfigs[i].mode == OUTPUT) || (gpioConfigs[i].mode == OUTPUT_OPEN_DRAIN)) {
+            if (doc.containsKey((const __FlashStringHelper*)gpioConfigs[i].name)) {
+                if (!doc[(const __FlashStringHelper*)gpioConfigs[i].name].is<unsigned int>()) {
```
nitpick: I think there's a macro to do this type of work, `F()` or something like that. If not, can you use `static_cast<const __FlashStringHelper *>()` instead?

hetii replied:
Changing it to static_cast/F trigger more work in around builder.py and generated configs, so I decide to skip that part, at least for now.

depau commented on src/ttyd.cpp:
```diff
@@ -9,14 +9,48 @@
+#include "xschedule.h"
 #include "ExtendedSerial.h"
 
-void TTY::shrinkBuffers() {
```
Why is `shrinkBuffers()` gone? Is it replaced by `TTY::end()`, which I see is called by `WiSeServer::end()`?

hetii replied:
The buffer size can only be modified when the uart is closed. You may even receive warnings about this.
This part of the code was only used during OTA or reboot, so I change it to be part of the end function.

depau commented on src/ttyd.cpp:
```diff
@@ -101,7 +135,7 @@ void TTY::sendWindowTitle(int64_t clientId) {
     size_t titleLen = 1 + snprintWindowTitle(windowTitle + 1, 99);
     AsyncWebSocketMessageBuffer *wsBuffer = websocket->makeBuffer((uint8_t *) windowTitle, titleLen);
-
+    if (!wsBuffer) return;
```
Did this actually happen? It sounds like a bug

EDIT: oh I see, does it happen with `sendInitialMessages` on ESP32?

hetii replied:
As with my previous comment about nullptr, I don't remember the details anymore, but it was probably added during stability testing.

depau commented on src/ttyd.cpp:
```diff
@@ -556,31 +622,84 @@ void TTY::dispatchUart() {
+    return gpioConfigs;
+}
+
+void TTY::sendGpioStates(char force) {
```
Note to self: I need to make sure `ttyc` at the very least safely ignores this message.

depau commented on src/ttyd.cpp:
```diff
@@ -293,19 +342,26 @@ void TTY::handleWebSocketPong(uint32_t clientId) {
 }
 
 void TTY::broadcastBufferToClients(AsyncWebSocketMessageBuffer *wsBuffer) {
+    if (!wsBuffer) return;
```
IDK, ignoring this makes me feel like we're ignoring bugs. How did you hit all these null pointers you've added checks for?

hetii replied:
All these nullptr checks were added during testing when esp restarted and the browser tried to immediately reconnect to the websocket, which sometimes led to further crashes.

I don't remember the details of each nullptr workaround, but the overall goal was to prevent the ESP from restarting.

I decided that if something didn't work on the ESP side, refreshing the page with wi-se would be less harmful than self restarting and sending random bytes to the target over uart.

depau commented on src/ttyd.cpp:
```diff
@@ -293,19 +342,26 @@ void TTY::handleWebSocketPong(uint32_t clientId) {
         // Fast no-copy path
         websocket->binaryAll(wsBuffer);
     } else {
+#ifdef LEGACY_LIB
         wsBuffer->lock();
+#endif
         for (int i = 0; i < wsClientsLen; i++) {
             AsyncWebSocketClient *client = websocket->client(wsClients[i]);
+            if (!client) continue;
             if (client->status() == WS_CONNECTED) {
                 client->binary(wsBuffer);
             }
         }
+#ifdef LEGACY_LIB
         wsBuffer->unlock();
         websocket->_cleanBuffers();
+#endif
```
nitpick: I really don't like the ifdefs; is the `LEGACY_LIB` flag caused by the fact that the library with my bugfixes is old and the API changed or is it the ESP32 version that doesn't have this?

hetii replied:
Adding LEGACY_LIB was necessary because I didn't know if your changes were still needed in this external library. Unfortunately, I didn't have time and knowledge to analyze the external lib in this regard, so I left the option to easily switch between the two versions.

depau commented on src/ttyd.cpp:
```diff
@@ -412,23 +469,31 @@ void TTY::performHousekeeping() {
 }
 
 bool TTY::wsCanSend() {
+    if (wsClientsLen == 0) {
+        return false;
+    }
```
potentially a bug: We definitely don't want to activate flow control when no client is connected, that would mess many things up on the device this is connected to.

Flow control should only be triggered if we're unable to keep up sending the data, in order to request the attached device to pause for a moment.

Is there any other subtle reason why you had to add this?

hetii replied:
As with nullptr, this is a double check to see if we still have clients on the line.
Harmless during normal operation.

depau commented on builder.py:
```diff
@@ -34,10 +34,17 @@
 }
 
 
+class SafeDict(dict):
+    def get(self, key, default=None):
+        val = super().get(key, default)
+        return default if val is None else val
```
Isn't this exactly what `dict.get()` does?

hetii replied:
No, dict.get() check key and I check here the value.

depau commented on builder.py:
```diff
@@ -212,6 +231,36 @@ def LED_ON_TIME(self):
     def LED_OFF_TIME(self):
         return self.jq('.board.leds.off_time', 15)
 
+    @property
+    def TARGET_GPIO_INITS(self):
+        jq = self.jq('.board.target', [])
+        l = []
+        for i,v in enumerate(jq):
+            v = SafeDict(v)
+            gpio = v['gpio']
+            mode = v.get('mode','INPUT')
+            init = 'TARGET_GPIO_ONINIT' if v.get('init', False) else 'TARGET_GPIO_LOCKED'
+            inverted = str(v.get('inverted','false')).lower()
+            l.append("{%s, %s, %s, %s, 0, &gpio_%s_dval, gpio_%s_name, gpio_%s_desc, gpio_%s_color}" % (gpio, mode, init, inverted, i, i, i, i))
+        return '\\\n%s\n' % ', \\\n'.join(l) if l else ''
+
+    @property
+    def TARGET_GPIO_STRINGS(self):
+        jq = self.jq('.board.target', [])
+        l = []
+        for i,v in enumerate(jq):
+            v = SafeDict(v)
+            l.append("const uint64_t gpio_%s_dval PROGMEM = %d;" % (i, v.get('dval', 0)))
+            l.append("const char gpio_%s_name[] PROGMEM = \"%s\";" % (i, v.get('name', "gpio%s" % v['gpio'])))
+            l.append("const char gpio_%s_desc[] PROGMEM = \"%s\";" % (i, v.get('desc', "GPIO %s" % v['gpio'])))
+            l.append("const char gpio_%s_color[] PROGMEM = \"%s\";" % (i, v.get('color', "#333" if v.get('mode','') == 'OUTPUT' else "limegreen")))
+        return '\\\n%s' % ' \\\n'.join(l) if l else ''
+
+    @property
+    def TARGET_GPIO_COUNT(self):
+        jq = self.jq('.board.target', [])
+        return '%s' % len(jq)
```
nitpick: Can you please use f-strings here instead of `%`?

hetii replied:
Will do.

depau commented on README.md:
```diff
@@ -6,19 +6,17 @@
 
-This software allows for using an ESP8266 board as a remote UART terminal. It is very fast, reaching (with [caveats](#caveats)) up to
-1500000bps rates.
+This software allows you to use the ESP8266 board as a remote UART terminal with GPIO port control and reading.
```
Now ESP32 too!

nit: `the ESP8266 board` → `an ESP8266 or ESP32 board`

depau commented on README.md:
```diff
@@ -6,19 +6,17 @@
-With some changes it may work with ESP32 but there's no interest for that at the moment. Pull requests are welcome.
+It is intended as a firmware for the [Wi-Se](https://github.com/Depau/wi-se-hw/)
+boards (which you can order and build at your favorite PCB manufacturer),  
+though it will work just fine with a normal ESP8266 breakout board.
```
ESP32 too :)

hetii replied:
ok

depau commented on README.md:
```diff
@@ -166,25 +238,78 @@ tty | grep -qE '/dev/tty[A-Za-z]+[0-9]*' && stty ixon ixoff
+
+If you're using the web client, this can close the browser, depending on which one you use.
+
+Chrome browser seams to ignore `Ctrl`+`Q` and pass it like it is, at least Linux version.
```
typo: `seams` → `seems`

depau commented on README.md:
```diff
@@ -166,25 +238,78 @@ tty | grep -qE '/dev/tty[A-Za-z]+[0-9]*' && stty ixon ixoff
 
-Another possible reason is that the firmware crashed. When the firmware crashes, the UART goes out of control until execution restart. This
-may result in sending a "break condition", which causes [`agetty`](https://man.archlinux.org/man/agetty.8) to switch to the next baud rate.
+Current version of wi-se web client will send `Ctrl`+`Q` on each press of `Ctrl`+`` ` ``to unlock your terminal.
```
nitpick: This is amazing! I can spot one potential problem though: the backtick `` ` `` may not be easy to type on some keyboards. For instance I normally use the US layout, but I'm from Italy and on the Italian layout `` ` ``, as well as `~`, is not even printed on the keyboard.

I was going to tell you if you could do something like `Ctrl`+`|` or `Ctrl`+`\`, which are probably not used by applications and should be available on all keyboards, though on a second thought I think, if possible, it could also be a nice idea to implement the [`ttyc` chords](https://github.com/depau/ttyc/blob/main/cmd/ttyc/handlers/stdfds.go#L25-L69) (`Ctrl`+`T` followed by another keypress) on the web front-end, adding one to explicitly send `xoff`. What do you think?

hetii replied:
Yes, we can add any mappings here.

Check how I did it in:
https://github.com/hetii/wi-se-sw/blob/072807c664b9bd041a78e5ec8cf765ef0d683c5c/html/src/components/terminal/index.tsx#L406

The point is that we need to be sure that the shortcut we use works without any problems in different web browsers, operating systems, and keyboards layouts.

Maybe let's consider that as a next step?

depau reviewed (changes requested):
Code review done. Great changes, thank you very much! I did spot one potential bug regarding flow control, all the other issues are mostly just nitpicks.

One thing worries me though, I'll state this here since I saw very many instances. I like the new null checks, but I'm a bit worried ignoring them is not the right thing to do. Usually when stuff returns nullptr it's a sign of a bug, so are you sure we shouldn't be crashing instead?

hetii committed:
Fix review issues.
